### PR TITLE
u2o: add policy route for u2o when disabling lb

### LINKED ISSUE #3814: u2o: underlay pods can‘t visit k8s service when disabling lb
# Bug Report

关闭 LB 功能并开启 Underlay 网络的 u2oInterconnection 后，Underlay 网络下的 Pod 能够访问 ovn-default 下的 Pod，但不能访问 cluster ip 映射的 Pod 服务

## Expected Behavior

开启 LB 能直接访问 service 的 cluster ip，期望关闭 LB 也能访问 service 的 cluster ip

## Actual Behavior

关闭 LB 后不支持访问 service 的 cluster ip

## Steps to Reproduce the Problem

1.修改 install.sh 脚本，修改 `ENABLE_LB=${ENABLE_LB:-true}` 为 `ENABLE_LB=${ENABLE_LB:-false}`
2.创建 Underlay subnet 并开启 u2o
3.创建 service 映射 ovn-default 下的 Pod 服务，并在 Underlay 网络下的 Pod 访问 service 的 cluster ip

## Additional Info

- Kubernetes version:

  **Output of `kubectl version`:**

  ```bash
  v1.25.3

  ```

- kube-ovn version:
  
  ```bash
  1.13.0

  ```

- operation-system/kernel version:
  
  **Output of `awk -F '=' '/PRETTY_NAME/ { print $2 }' /etc/os-release`:**
  **Output of `uname -r`:**

  ```bash
  Ubuntu 22.04.4 LTS
  6.5.0-21-generic
  ```

1、开启 LB 后，Undeylay Pod 是通过 ovn 的 lb 将访问 service 的流量转发到 ovn-default的
2、关闭 LB 后，由于 ovn0 能够访问默认 vpc 下的service，考虑通过路由将访问 service cluster ip 的流量 reroute 到宿主机的 ovn0 上，借助 ovn0 实现对 service 的访问。
```shell
 kubectl ko nbctl lr-policy-list ovn-cluster
Routing Policies
     29400                 ip4.src == $vlan.master_ip4         reroute                100.64.0.2
     29400                 ip4.src == $vlan.worker_ip4         reroute                100.64.0.3
```

### PR DESCRIPTION
**Pull request by zcq98**
# Pull Request

- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

## What type of this PR

Examples of user facing changes:

- Features
- Bug fixes

## Which issue(s) this PR fixes

Fixes #3814

### CODE REVIEW

File: pkg/controller/subnet.go [hunk 1]
```diff
@@ -1721,11 +1721,47 @@ func (c *Controller) reconcileOvnDefaultVpcRoute(subnet *kubeovnv1.Subnet) error
...
+						return err
+					}
+				}
+			}
 		} else {
 			if err := c.deletePolicyRouteForU2OInterconn(subnet); err != nil {
 				klog.Errorf("failed to delete policy route for underlay to overlay subnet interconnection %s, %v", subnet.Name, err)
 				return err
 			}
+			if !c.config.EnableLb {
```
Comment (bobz965) [inline pkg/controller/subnet.go:1751]: 考虑两个 if !c.config.EnableLb { 合并逻辑

File: pkg/controller/subnet.go [hunk 2]
```diff
@@ -1721,11 +1721,47 @@ func (c *Controller) reconcileOvnDefaultVpcRoute(subnet *kubeovnv1.Subnet) error
...
+				pgs, err := c.OVNNbClient.ListPortGroups(map[string]string{logicalRouterKey: subnet.Spec.Vpc, logicalSwitchKey: subnet.Name, u2oKey: "true"})
+				if err != nil {
+					klog.Errorf("failed to list u2o port groups with u2oKey is true for subnet %s: %v", subnet.Name, err)
+					return err
+				}
+				for _, pg := range pgs {
+					klog.Infof("delete u2o port group %s for subnet %s", pg.Name, subnet.Name)
```
Comment (bobz965) [inline pkg/controller/subnet.go:1758]:
```suggestion
					klog.Infof("failed to delete u2o port group %s for subnet %s", pg.Name, subnet.Name)
```

File: pkg/controller/subnet.go [hunk 3]
```diff
@@ -1919,6 +1955,67 @@ func (c *Controller) reconcileU2OInterconnectionIP(subnet *kubeovnv1.Subnet) err
 
 			needCalcIP = true
 		}
+
+		// create port groups for policy route to visit k8s service when enabling U2OInterconnection and disabling lb
+		if subnet.Spec.Vpc == c.config.ClusterRouter && !c.config.EnableLb {
```
Comment (bobz965) [inline pkg/controller/subnet.go:1960]:
```suggestion
		if  !c.config.EnableLb && subnet.Spec.Vpc == c.config.ClusterRouter {
```

File: pkg/controller/subnet.go (hunk 2 above)
Comment (zcq98) [reply pkg/controller/subnet.go:1758]: 这个日志是开始删除 port group，不是删除失败

**Comment by zhangzujian** [comment]
There is another `enableLb` filed in subnet definition: `.spec.enableLb`.

**Comment by zcq98** [comment]
> There is another `enableLb` filed in subnet definition: `.spec.enableLb`.

1、关闭全局 LB，子网级别 LB 会失效，此时需要下发策略路由保证开启 U2O 后 Underlay Pod 能访问 Service
2、开启全局 LB，若子网关闭了自身的 LB，则仍会出现开启 U2O 后无法访问 Service 的问题
因此考虑在开启 U2O 后无论是否开启 LB 都下发策略路由，当 LB 开启后，访问 Service 走 LB，不会再匹配策略路由，关闭 LB 时，访问 Service 走策略路由，经 ovn0 访问 Service

File: pkg/controller/subnet.go [hunk 4]
```diff
@@ -1919,6 +1951,67 @@ func (c *Controller) reconcileU2OInterconnectionIP(subnet *kubeovnv1.Subnet) err
 
 			needCalcIP = true
 		}
+
+		// create port groups for policy route to visit k8s service when enabling U2OInterconnection and disabling lb
+		if !c.config.EnableLb && subnet.Spec.Vpc == c.config.ClusterRouter {
```
Comment (bobz965) [inline pkg/controller/subnet.go:1956]: 判断 u2o

**Review by bobz965 — CHANGES_REQUESTED** [review]

File: pkg/controller/subnet.go (hunk 4 above)
Comment (zcq98) [reply pkg/controller/subnet.go]: 这里的逻辑已经在 u2o 开启的里了，只需要把 !c.config.EnableLb 去掉就行

Comment (zcq98) [reply pkg/controller/subnet.go]: 已修正

**Review by bobz965 — APPROVED** [review]

**Comment by zhangzujian** [comment]
We should check whether ovn lb is enabled or not before creating the logical router policy.

**Comment by zcq98** [comment]
> We should check whether ovn lb is enabled or not before creating the logical router policy.

Service 的 backends 不一定是 ovn-default 下的 Pods，来一个 Service 判断一下 backends 所在的子网是否开启  LB 似乎不太合适

**Comment by zcq98** [comment]
hi @zhangzujian，could you help review this？

**Comment by bobz965** [comment]
> We should check whether ovn lb is enabled or not before creating the logical router policy.

it looks already done.

**Pull request merged**